### ISSUE TITLE
ToolbarAndroid removed in 0.61

### ISSUE THREAD

**Issue by thomasvm** [facebook/react-native#26591]
React Native 0.61 removed `ToolbarAndroid `from the provided components. This is not mentioned in the release docs and the component was not extracted into community package either. So there is no clear "path forward". Also, ToolbarAndroid is still mentioned as a provided components in the latest docs.

https://facebook.github.io/react-native/docs/toolbarandroid

React Native version: 0.61
System:
    OS: Windows 10
    CPU: (4) x64 Intel(R) Core(TM) i5-6200U CPU @ 2.30GHz
    Memory: 3.51 GB / 15.88 GB
  Binaries:
    Node: 10.11.0 - C:\Program Files\nodejs\node.EXE
    Yarn: 1.12.1 - ~\AppData\Roaming\npm\yarn.CMD
    npm: 6.9.0 - C:\Program Files\nodejs\npm.CMD
  IDEs:
    Android Studio: Version  3.5.0.0 AI-191.8026.42.35.5791312

## Steps To Reproduce

1. Upgrade project to react-native 0.61
2. Code that is using `ToolbarAndroid ` fails with error message  "Invariant Violation: Element type is invalid: expected a string (for built-in components) or a class/funcction"

**Comment by Smoothsmith** [comment]
I found the commit that removes it:
https://github.com/facebook/react-native/commit/31ab94770311f56182734b57c48ffcd765096dc5#diff-39808e06333026f9ddb02287226012e1

But I don't know how to use that information to access/continue using it.

**Comment by thomasvm** [comment]
@Smoothsmith There is another commit that removes the Java part to FB internal
https://github.com/facebook/react-native/commit/cbbbb455dda313cca1a1763cc5453079ce87bc73

What I ended up doing - and what I guess will be your best shot at the moment - is to include the JavaScript and Java code in our own codebase (with adjusted namespaces etc) and consume it from there.

**Comment by stefanhuber** [comment]
This is so painful... Is there any option to create a simple toolbar in Android without putting some native code somewhere?

**Comment by forki** [comment]
@cpojer Is there an replacement? It's still in the docs and hit uns really hard

**Comment by connectdotz** [comment]
hmmm... this seems like a major API breakage without any mention in changelog or blog, makes one wonder if this is an intentional change 🤔 ? We are already at 0.61.5 and there is still no official guidance on this issue... does the react-native core team not care about the tooling community or adoption of the new release? It would be great to have at least a discussion...

**Comment by Naturalclar** [comment]
@connectdotz I may be wrong, but I think removal of ToolbarAndroid has been discussed as part of a [Lean Core](https://github.com/facebook/react-native/issues/23313), and I believe it was considered deprecated since RN 0.59

**Comment by forki** [comment]
@Naturalclar it's not in the release notes. but it's still in the docs. see https://facebook.github.io/react-native/docs/toolbarandroid

**Comment by connectdotz** [comment]
I stand corrected, @Naturalclar thanks for the link, it provided the missing context for me. 

According to the lastest [comment](https://github.com/facebook/react-native/issues/23313#issuecomment-541481129) from @cpojer on Oct 13, 2019, which he announced the removal is on hold with the sound reason. I completely agree with his approach: move these "unnecessary core components" to external/optional packages first, then deprecate the component. This is clearly demonstrated for components like [Slider](https://facebook.github.io/react-native/docs/slider), where there is clear guidance for migration meanwhile the component is still in the react-native repo.

However, this is not the case for `ToolbarAndroid`... I am curious why wouldn't this component be restored until it has an external package alternative, like any other to-be-deprecated components... 🤔

**Comment by connectdotz** [comment]
I stand corrected, @Naturalclar thanks for the link, it provided the missing context for me, but it only further deepened the puzzle of **how** ToolbarAndroid was removed...

After reading PR #24999, which enabled the removal of ToolbarAndroid, it became more clear how ToolbarAndroid could be replaced. But the migration is not clearly documented, to say the least... I understand this is just a "convenient" class and there is nothing special but a composite of core building blocks... However, the same thing can be said for many of the to-be-deprecated classes in the [lean core](https://github.com/facebook/react-native/issues/23313) initiative... again, I am not sure why `ToolbarAndroid` is being handled differently, i.e. it was removed without any clear documentation on migration path nor grace period for 3rd party libraries and user applications to adapt?

Will @cpojer and @ericlewis consider updating the [document](https://facebook.github.io/react-native/docs/toolbarandroid) to clarify the migration approach? Better yet, maybe even bring back the `ToolbarAndroid` and deprecate it later with the rest of the lean core classes?

**Comment by ericlewis** [comment]
One could reimplement it as it’s own module based on what was removed. 

This was quite a while ago. But if I recall the reasoning correctly, it was that it’s been long deprecated with the idea that more complete navigation solutions better implement this sort of thing.

**Comment by connectdotz** [comment]
@ericlewis I completely agree that these kind of convenient components should live outside of react-native core, just that the approach for ToolbarAndroid is inconsistent with the rest of lean-core initiative components, not to mention the migration barrier it has imposed: first, you will get puzzling compile error, but nothing is mentioned in changelog, release note, and the official document didn't say anything should be different... well, you got the picture...

Furthermore, If ToolbarAndriod is a common component, I suspect it is, instead of everybody duplicates the code you proposed, it would probably be better to create an external package, as many of the lean-core components moving toward, no?

**Comment by ericlewis** [comment]
I don’t recall it being rushed, since it was removed probably close to a year ago and the discussion about it being public. 

Anyone from the community is free to implement their own version of this and provide it as open source. 

They are also free to open pull requests to update documentation. 

I’d encourage doing both things!

**Comment by forki** [comment]
@ericlewis the problem is not that it is removed. The problem is that it's missing in the release notes and the docs still mentions it as supported

**Comment by TheSavior** [comment]
The changelog generator probably skipped this commit because the title includes the word “internal”. This should definitely be in the changelog. Can someone send a PR to the changelog please?

**Comment by cpojer** [comment]
At the time it was my understanding that ToolbarAndroid had been deprecated for a while and was a rarely used component that can be implemented purely in JavaScript using styles. That's why I decided to remove it altogether without asking others to help maintain it as a separate new module.

If this is indeed a module that people still heavily rely on, here are some next steps:
* Copy the source to a new repository and make it into a third-party repo OR provide a new solution using JavaScript only that mirrors the same functionality.
* Update the changelog which missed this removal (sorry about that, as Eli says it was likely filtered).
* Remove the existing documentation from the website.

**Comment by forki** [comment]
"purely Javascript"? Are you sure?

**Comment by cpojer** [comment]
Here is an example implementation by @ericlewis which we are using in RNTester: https://github.com/facebook/react-native/pull/24999

**Comment by forki** [comment]
But that's not the native control, right? And it's also not showing ActionIcons and how ActionIcons overflow.

**Comment by forki** [comment]
@thomasvm any chance you can open source that part that you have extracted?

**Comment by connectdotz** [comment]
thanks, @cpojer for swift actions. 

I wonder, instead of removing the document, will it be more helpful to state at the top that it is deprecated since 0.61(?) and suggest a way forward for both older and current version users? 

Maybe I am still missing some context... was`ToolbarAndroid` deprecated a long time ago?  I bump into this issue with a new react-native app (v0.61.5) with `react-native-vector-icons` and `react-native-gesture-handler`, both have `ToolbarAndriod` dependencies. These 3rd party libraries are quite popular among react-native communities. Judging they all still have dependencies on this class (react-native-gesture-handler did merge the removal PR just 4 days ago but haven't released the npm package yet), I got the impression this is a fairly new issue(?)...

@forki observation is revealing that perhaps a pure javascript solution is not quite fully functional-equivalent yet? Maybe a little bit more time for the new solution to be fully tested and mature, in a 3rd party library before deprecating the `ToolbarAndroid` would be safer?

On the other hand, if this is indeed a rarely used component, then one can argue all 3rd party libraries should just remove it rather than restoring it in the core library during the transition. I don't know how to determine ToolbarAndroid popularity on behave of the 3rd party libraries and the communities, but nevertheless it is best to give the library/app owners some warning and time to decide how to support, or not, about this component?

**Comment by cpojer** [comment]
@connectdotz I definitely dropped the ball on this one a little bit, and I'm sorry.

Regarding gesture-handler, I did work with @kmagiera to get the project ready. https://github.com/software-mansion/react-native-gesture-handler/pull/617 was the result of that back at the same time when ToolbarAndroid was removed so that version of the gesture handler should work with both the old and new version of React Native.

We will not be restoring this within react-native so the recommendation is to make a new repository with this functionality.

**Comment by cesardeazevedo** [comment]
Anyone already started moving ToolbarAndroid to a new repository? My library [react-native-collapsing-toolbar](https://github.com/cesardeazevedo/react-native-collapsing-toolbar) highly relied on the native implementation.

**Comment by connectdotz** [comment]
@cpojer thanks for the explanation, no apology needed, just trying to understand the whole story so we can better plot the path forward.

With the information you mentioned above, I was able to draw a more clear picture now. It seems the compile error started to be reported after necolas/react-native-web#512, which removed `ToolbarAndroid` support around 18 days ago for `react-native-web@0.12.0-rc.x`. Apps with the latest versions (react-native@0.61.x, react-native-web@0.12.0-rc.x)  started to see compile error, even without directly using `ToolbarAndroid`. This is because 3rd party libraries are linked against users' own react-native library in the react-native-web paradigm, which the regular react-native apps won't suffer unless they use `ToolbarAndroid` directly...

I understand the decision not to restore `ToolbarAndroid`, that's fine, then we should get the external library established ASAP, can you create the `react-native-community/toolbar-android` repo and maybe move the original ToolbarAndriod code over? Hmmm... actually, I wonder wouldn't it be more consistent and speedy to just take the whole to-be-deprecated components into its own monorepo package, something like `lean-components`, which should have a clear mandate to have only the react and react-native dependency, distinguished from all the other 3rd party libraries out there? This way you guys can speed up the lean core initiative progress without blocked by community ownership as well as providing a more consistent structure/process around these components 🤔 ... 

Anyway, for people who landed here looking for surviving react-native-web and ToolbarAndroid deprecation, my current workaround is to downgrade react-native-web to `^0.11`:  it appeared to avoid the compile error, as long as you don't use `ToolbarAndroid` directly. Hopefully, we will have a comprehensive solution soon 🤞

**Comment by cpojer** [comment]
@connectdotz I created https://github.com/react-native-community/toolbar-android and gave you access. Feel free to send PRs there, merge them and publish a new package.

**Comment by connectdotz** [comment]
@cpojer that was clever 😏 ok, I will shut up and actually **do** something to resolve this issue 😊 

I will take a shot at seeding the repo with the original `ToolbarAndriod`. We can maybe look to include @ericlewis's alternative js implementation in the following release...

**Comment by ericlewis** [comment]
@connectdotz You know what, I was a bit confused by the confusion until @cpojer chimed in. I did indeed create a pure JS version of this which completely mimicked the abilities needed by the RNTester app (as well as tests). I think it might live there? Either way, sorry for the inconvenience!

**Comment by ericlewis** [comment]
We somehow dropped the ball here, I would assure you that you can definitely achieve fantastic results with a pure javascript solution!

**Comment by forki** [comment]
From a user's perspective there were absolutely no signs that this standard Android component is not treated as something properly supported. There was no such hint in the docs. There are even fixes documented in the release notes.
Regarding full js solution: I think the currently proposed way is a good way forward. Extraction of the deleted code into a react-native-community control will unblock people. Afterwards it may or may not be improved.

**Comment by connectdotz** [comment]
@cpojer I am not able to add file... no permission to commit directly; and github won't let me fork https://github.com/react-native-community/toolbar-android either because it is empty... maybe adding a dummy file like README.md? Thanks.

**Comment by connectdotz** [comment]
@ericlewis 

> I did indeed create a pure JS version of this which completely mimicked the abilities needed by the RNTester app (as well as tests). I think it might live there?

I think we can definitely host both implementations in the repo after the initial release (with the extracted code). The new pure js implementation can be named something like`ToolbarAndroidJS` so people can experiment with it safely until it is mature. 

> I am pretty sure this was never a proper supported component

Do you care to elaborate?

**Comment by connectdotz** [comment]
ok, initial check-in is done. However, I only tested it with RN 0.60 and 0.61.5 (see examples there), which went smoothly with autolink. I don't have older RN repo to test with the manual setup, if you do please try it and see if we need to update the README.md or something else...

I have check in the npm package `react-native-community-toolbar-android-0.1.0.tgz` in the repo but haven't published yet, thinking it surely needs a bit more testing than the examples there... 

You can install the package from the tgz file, for example `yarn add file:react-native-community-toolbar-android-0.1.0.tgz`. Feel free to file issues in https://github.com/react-native-community/toolbar-android  so we can get the package ready ASAP. Thanks.

**Comment by connectdotz** [comment]
@cpojer [circleci](https://circleci.com/add-projects/gh/react-native-community) won't let me add `@react-native-community/toolbar-android`, do you mind adding it for us? thx.

**Comment by cpojer** [comment]
Should be all set :)

**Comment by khizerism** [comment]
As im new to react-native, this is the best solution i found so far, hope it helps. 

https://github.com/react-native-community/toolbar-android

First, add npm install @react-native-community/toolbar-android --save or yarn lover add yarn add @react-native-community/toolbar-android

Second, Run react-native start command.

Third, Run react-native run-android

Happy coding :)

**Comment by thomasvm** [comment]
Since ToolbarAndroid is now extracted into a separate library and now that the documentation mentions this, I am going to close this issue. Thanks for the work everybody!
